Normative: tighten language in non-ISO-calendar mergeFields()

ptomato opened the pull request:
Note, DefaultMergeCalendarFields still exists in this branch but one use of it would be removed in #2310, and #1418 would consolidate this definition of Temporal.Calendar.prototype.mergeFields() with the earlier one in the 262 part of the proposal. So ultimately, this would be the only definition of mergeFields.

To do, if we decide this is the right approach:
- Check that all the places that might produce abrupt completions are correct
- Change the reference code to follow this spec

Closes: #2407

justingrant commented on spec/intl.html:
```diff
@@ -2123,14 +2106,26 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
+                1. Else if _key_ is *"monthCode"*, let _excludedKeys_ be « *"month"* ».
+                1. Else let _excludedKeys_ be « ».
+              1. Else,
+                1. Let _excludedKeys_ be the List of property keys that are mutually excluded by _key_.
```
Do you want to have a bit more context about what "mutually excluded" means? The original text was IMO a bit clearer about this.

ljharb replied:
+1, i really have no idea how i'd implement this line. can we use explicit algorithm steps?

ptomato replied:
We could pull this part out into a separate op or whatever, but I'd like to resolve the discussion in #2407 whether this is the approach we want before doing any fine-tuning of the language in this PR.

ptomato replied:
I've pulled the calendar-dependent part out into a separate operation with an explanation of "mutually excluded" and examples.

gibson042 commented on spec/intl.html:
```diff
@@ -2123,14 +2106,26 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
+              1. If _calendar_.[[Identifier]] is *"iso8601"*, then
+                1. If _key_ is *"month"*, let _excludedKeys_ be « *"monthCode"* ».
+                1. Else if _key_ is *"monthCode"*, let _excludedKeys_ be « *"month"* ».
+                1. Else let _excludedKeys_ be « ».
```
I don't think this works; you probably want to define _excludedKeys_ outside the loop and non-destructively update it in the loop body.

gibson042 commented on spec/intl.html:
```diff
@@ -2123,14 +2106,26 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
+                1. NOTE: For example, in calendars that use eras, if _key_ is *"year"*, let _excludedKeys_ be « *"era"*, *"eraYear"* ». _excludedKeys_ may be empty.
+                1. NOTE: In calendars where eras do not start and end at year boundaries, note that _excludedKeys_ should contain *"era"* and *"eraYear"* if _key_ is *"day"*, *"month"*, or *"monthCode"*.
+              1. For each element _excludedKey_ of _excludedKeys_, do
+                1. Perform ! DeletePropertyOrThrow(_merged_, _excludedKey_).
```
It would be nice if this did not produce observable property reordering.

justingrant commented on spec/intl.html:
```diff
@@ -2139,13 +2182,22 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
+                1. If _key_ is *"month"*, append *"monthCode"* to _excludedKeys_.
+                1. Else if _key_ is *"monthCode"*, append *"month"* to _excludedKeys_.
+              1. Else,
+                1. Set _excludedKeys_ to the list-concatenation of _excludedKeys_ and CalendarMutuallyExclusiveKeys(_calendar_, _key_).
```
Is it OK that `CalendarMutuallyExclusiveKeys` can return the same property multiple times? For example, if `_additionalFields_` includes both `era` and `eraYear`, then `year` will be added twice to `_excludedKeys_`.  And for Japanese, if `_additionalFields_` includes both `day` and `year` then I think `era` and `eraYear` will be present 3x in `_excludedKeys_`.

ptomato replied:
I guess there's a [Set specification type](https://tc39.es/ecma262/#sec-set-and-relation-specification-type) we can use for this.

justingrant commented on spec/intl.html:
```diff
@@ -1676,20 +1676,63 @@ <h1>
+          </emu-table>
           <emu-note>
-            For example, if _fields_ contains *"year"* but _additionalFields_ contains *"era"* and *"eraYear"*, then the returned list must not include *"year"* in order to avoid ambiguity or conflict between different ways of specifying the year.
+            In calendars where eras do not start and end at year and month boundaries, note that the returned List should contain *"era"* and *"eraYear"* if _key_ is *"day"*, *"month"*, or *"monthCode"*, in addition to *"year"*, because it's possible for changing the day or month to cause a conflict with the era.
```
```suggestion
            In calendars like Japanese where eras do not start and end at year and/or month boundaries, note that the returned List should contain *"era"* and *"eraYear"* if _key_ is *"day"*, *"month"*, or *"monthCode"*, in addition to *"year"*, because it's possible for changing the day or month to cause a conflict with the era.
```

justingrant commented on spec/intl.html:
```diff
@@ -2139,13 +2182,22 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
             1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
```
In a post-#2482 world, does this Calendar method get called when a string id was used for the calendar? Or is there some AO that's called instead?

I'm asking because looking at the spec, Temporal will never call `mergeFields` with either `era` or `eraYear` present in `fields`. So "the returned List should contain *"era"* and *"eraYear"*" is a no-op unless userland code is calling `mergeFields`. Is this correct?

If so, then should there be a simpler version of this implementation that's used for non-observable merge-fields operations?  Or does that already exist?

ptomato commented on spec/intl.html:
```diff
@@ -1676,20 +1676,63 @@ <h1>
+              <tbody>
+                <tr>
+                  <td>*"era"*</td>
+                  <td>« *"year*" »</td>
```
As per https://github.com/tc39/proposal-temporal/issues/2465#issuecomment-1402211524 we need to exclude eraYear here too.

ptomato commented:
Un-drafting, because this change achieved consensus at the TC39 plenary meeting of 2023-01-31.

I've changed the reference implementation to work more like the spec (cc @justingrant), and take the updated order of operations into account. I've also added a bug fix for our patched CopyDataProperties so that it matches the spec (cc @gibson042).

Tests are in https://github.com/tc39/test262/pull/3778.

justingrant commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -269,12 +269,13 @@ export const ES = ObjectAssign({}, ES2022, {
       });
+      if (excluded) return;
 
       var enumerable =
```
Since we're changing this function anyways, should we replace var with let or const?

ptomato replied:
Richard wanted to keep it the same as the original CopyDataProperties as much as possible.

justingrant replied:
OK

justingrant commented on spec/intl.html:
```diff
@@ -1685,20 +1685,64 @@ <h1>
+            </dd>
           </dl>
+          <p>
+            Mutually exclusive means that the fields are alternate, potentially conflicting, ways of expressing the same data.
```
Is this sentence (and the name of this AO) accurate if `day` causes exclusion of `era` and `eraYear` in Japanese?  `day` isn't an alternate way to specify the year, but it's returned nontheless.

ptomato replied:
This is a good point. I think renaming it to CalendarKeysToExclude and adjusting the description a bit should be sufficient.

justingrant reviewed (approved):
Looks good. Minor comments only.

gibson042 commented on spec/intl.html:
```diff
@@ -2148,13 +2192,24 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
             1. Perform ? CopyDataProperties(_additionalFieldsCopy_, _additionalFields_, « », « *undefined* »).
-            1. Return CalendarDateMergeFields(_calendar_.[[Identifier]], _fieldsCopy_, _additionalFieldsCopy_).
+            1. NOTE: Every property of _additionalFieldsCopy_ is a data property with non-*undefined* value, but some property keys may be Symbols.
+            1. Let _excludedKeys_ be an empty Set.
+            1. Let _additionalKeys_ be ! _additionalFieldsCopy_.[[OwnPropertyKeys]]().
+            1. For each element _key_ of _additionalKeys_, do
+              1. Add _key_ to _excludedKeys_.
+              1. If _calendar_.[[Identifier]] is *"iso8601"*, then
+                1. If _key_ is *"month"*, add *"monthCode"* to _excludedKeys_.
+                1. Else if _key_ is *"monthCode"*, add *"month"* to _excludedKeys_.
+              1. Else,
+                1. Let _toExclude_ be CalendarKeysToExclude(_calendar_, _key_).
+                1. Set _excludedKeys_ to the union of _excludedKeys_ and the set of _toExclude_'s elements.
+            1. Let _merged_ be OrdinaryObjectCreate(%Object.prototype%).
+            1. Let _excludedKeysList_ be the List of _excludedKeys_' elements.
+            1. Perform ? CopyDataProperties(_merged_, _fields_, _excludedKeysList_, « *undefined* »).
+            1. Perform ! CopyDataProperties(_merged_, _additionalFieldsCopy_, « »).
```
I think these steps would be more comprehensible if they were holistic rather than field-by-field, and also if _excludedKeys_ were renamed. Suggestion:
```suggestion
            1. If _calendar_.[[Identifier]] is *"iso8601"*, then
              1. Let _overriddenKeys_ be ISOFieldKeysToOverride(_additionalKeys_).
            1. Else,
              1. Let _overriddenKeys_ be CalendarFieldKeysToOverride(_calendar_, _additionalKeys_).
            1. Let _merged_ be OrdinaryObjectCreate(%Object.prototype%).
            1. Perform ? CopyDataProperties(_merged_, _fields_, _overriddenKeys_, « *undefined* »).
            1. Perform ! CopyDataProperties(_merged_, _additionalFieldsCopy_, « »).
```

It would also be nice if property order aligned with `{ ...fields, ...additionalFields }` (here and in DefaultMergeCalendarFields, of course):
```suggestion
            1. If _calendar_.[[Identifier]] is *"iso8601"*, then
              1. Let _overriddenKeys_ be ISOFieldKeysToOverride(_additionalKeys_).
            1. Else,
              1. Let _overriddenKeys_ be CalendarFieldKeysToOverride(_calendar_, _additionalKeys_).
            1. NOTE: The following steps ensure that property iteration order of _merged_ matches that of _fields_ as modified by omitting overridden keys and appending non-overlapping keys from _additionalFields_ in iteration order.
            1. Let _fieldsKeys_ be ! _fields_.[[OwnPropertyKeys]]().
            1. For each element _key_ of _fieldsKeys_, do
              1. Let _propValue_ be *undefined*.
              1. If _overriddenKeys_ contains _key_,
                1. Set _propValue_ to ! Get(_additionalFieldsCopy_, _key_).
              1. Else,
                1. Let _desc_ be ? <emu-meta effects="user-code">_fields_.[[GetOwnProperty]]</emu-meta>(_key_).
                1. If _desc_ is not *undefined* and _desc_.[[Enumerable]] is *true*, then
                  1. Set _propValue_ to ? Get(_from_, _key_).
              1. If _propValue_ is not *undefined*, perform ! CreateDataPropertyOrThrow(_merged_, _key_, _propValue_).
            1. Perform ! CopyDataProperties(_merged_, _additionalFieldsCopy_, « »).
```

ptomato replied:
I've taken a look at how this would shake out. I could go either way. I think iterating and calling the implementation-defined operation once per key is probably closer to how it'd be implemented in an engine, and codifies the fact that keys must exclude themselves. On the other hand, calling the implementation-defined operation on the set of keys.

I don't feel strongly about matching `{ ...fields, ...additionalFields }` either — if the returned object has a null prototype it's not possible to polyfill it that way anyway.

gibson042 replied:
I believe you are mistaken, because the third argument in `CopyDataProperties(_merged_, _fieldsCopy_, _overriddenKeys_, « *undefined* »)` has the effect of omitting keys that are present in _fieldsCopy_ which the subsequent `CopyDataProperties(_merged_, _additionalFieldsCopy_, « »)` can add back in a different position. For a simple example, consider how this algorithm requires `Reflect.ownKeys(Temporal.Calendar.from("iso8601").mergeFields({ year: 2023, month: 2, day: 18 }, { month: 3 }))` to return `["year", "day", "month"]` rather than `["year", "month", "day"]`.

gibson042 replied:
#2504

gibson042 commented on spec/intl.html:
```diff
@@ -2148,13 +2192,24 @@ <h1>Temporal.Calendar.prototype.mergeFields ( _fields_, _additionalFields_ )</h1
+                1. Set _excludedKeys_ to the union of _excludedKeys_ and the set of _toExclude_'s elements.
+            1. Let _merged_ be OrdinaryObjectCreate(%Object.prototype%).
+            1. Let _excludedKeysList_ be the List of _excludedKeys_' elements.
+            1. Perform ? CopyDataProperties(_merged_, _fields_, _excludedKeysList_, « *undefined* »).
```
It's worth noting that this observably prevents reading "month"/"monthCode"/etc. properties from a user-provided _fields_ object when _additionalFields_ includes covering fields, in contrast with the normal pattern of "observably copy all fields". Should this instead follow the normal pattern and then unobservably remove overridden fields that didn't come from _additionalFields_?

(and likewise for DefaultMergeCalendarFields)

ptomato replied:
I thought that [your earlier comment](https://github.com/tc39/proposal-temporal/pull/2474#discussion_r1071804934) had asked for it to be done this way, but looking back at it I now can't understand how I thought that. Sorry.

I could go either way on this. I think it's fine to not read fields if we're already aware we're not going to use them, but reading them was the status quo before this PR, so preserving that would make the scope of this change smaller.

gibson042 commented on spec/intl.html:
```diff
@@ -1685,20 +1685,64 @@ <h1>
           </dl>
+          <p>
+            Exclusive fields means the set of fields that changing the field named _key_ can potentially conflict with or invalidate.
+            This operation is relevant for calendars which accept fields other than the standard set of ISO calendar fields, in order to implement the Temporal objects' `with()` methods, and `Temporal.Calendar.prototype.mergeFields()` in such a way that the result is free of ambiguity or conflicts.
```
```suggestion
            This operation is relevant for calendars which accept fields other than the standard set of ISO calendar fields, in order to implement `Temporal.Calendar.prototype.mergeFields()` in such a way that the result is free of ambiguity or conflicts.
```

ptomato replied:
I think including the phrase about `with()` here makes it clearer how this ultimately relates to user code.

ptomato commented:
I've updated this. I think there are two open questions which I'd like to resolve as quickly as possible:
- A calendar-specific operation that gives you keys to override for each input key, or for the full list of input keys at once? (you can see the difference in the last commit I've added to this branch)
- Match observable accesses to what you'd get if you implemented this with `{...fields, ...additionalFields}` or avoid observable reads if we know we won't use the value? The former will require a corresponding update to https://github.com/tc39/test262/pull/3778.

justingrant commented on polyfill/lib/calendar.mjs:
```diff
@@ -1688,6 +1695,7 @@ const helperJapanese = ObjectAssign(
     // The last 3 Japanese eras confusingly return only one character in the
     // default "short" era, so need to use the long format.
     eraLength: 'long',
+    erasBeginMidYear: true,
```
I know we found the problem in Japanese, but does this only affect Japanese?  Or do other calendars need this behavior too? Would it be harmful to make all calendars behave the same?

ptomato replied:
I believe the only calendar affected is Japanese, but this is part of the calendar-dependent part of the spec anyway, so it wouldn't be a normative change if we wanted to tweak it in our reference code (unless it was covered later in the era and month codes proposal).

justingrant commented on spec/intl.html:
```diff
@@ -1685,20 +1685,72 @@ <h1>
+        <emu-clause id="sec-temporal-calendarfieldkeystooverride" type="abstract operation">
           <h1>
-            CalendarDateMergeFields (
+            CalendarFieldKeysToOverride (
```
Minor bikeshedding thing: "override" doesn't necessarily seem like the right word to describe what's happening, because `addtionalFields` often doesn't contain the fields being "overridden". The goal of this AO is to figure out which fields should be removed from `fields` so that the result, when merged, won't be ambiguous. So maybe some other possible names could be one of the following, or some other clearer name?

* FieldKeysToRemove
* FieldKeysToIgnore
* AmbiguousFieldKeys
* ConflictingFieldKeys (or FieldKeysThatConflict)

ptomato replied:
Any of those are fine, but I'd like move towards ending the bikeshedding, and pick one that everybody can live with and stick with it. I've changed the name a few times already, and had to rephrase the accompanying prose each time.

justingrant replied:
Deal. I'd suggest `*FieldKeysToIgnore` but any of those are OK with me.

ptomato replied:
There were no other comments, so I'll move ahead with that. If we need to iterate further on the name or the phrasing, that can be done in a separate PR.
